### PR TITLE
Cdk sample

### PR DESCRIPTION
**Pull request by flo-mair**
CDK Sample Deployment

Description
-----------

This adds a CDK sample to deploy the connector including:
- MSK Cluster
- VPC, SG & IAM Roles
- Kafka Producer Application on ECS
- Kafka Consumer Application on ECS
- MSK Connect Plugin for EventBridge Connector
- MSK Connector (if deploymentMode = FULL)

Test Steps
-----------

Checklist:
----------

- [x ] I have tested my changes. No regression in existing tests.
- [ ] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -97,7 +97,7 @@ The following minimal configuration configures the connector with default values
 `"json-values-topic"` with record keys as `String` and `JSON` values (without schema), and sending events to the custom
 EventBridge event bus `"kafkabus"` in region `"us-east-1"`.
 
-```json5
```
Comment (embano1) [inline README.md:100]: Why the change :) ?

File: cdk/cdk.context.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
+  "acknowledged-issue-numbers": [
+    25714
```
Comment (embano1) [inline cdk/cdk.context.json:3]: What is this for?

File: cdk/README.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+- Amazon MSK Serverless Cluster
+- Amazon S3 Bucket
+- IAM Roles & Security Groups
+- VPC & related components
+- Custom Amazon EventBridge Bus
+- Amazon ECS cluster with 2 Services (Data Producer & Transaction Analyzer)
+- (Optional) Amazon MSK Connector
+
+To learn more about those components please read the blog on the [AWS Big Data Blog](https://aws.amazon.com).
```
Comment (baldawar) [inline cdk/README.md:19]: Is there a better resource we can link ? This one seems too broad to be helpful.

File: cdk/README.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+cdk deploy --context deploymentMode=FULL
+```
+
+To deploy only the Prerequisites and later deploy the connector manually. Please run the following command in this directory:
+
+```shell
+cdk deploy --context deploymentMode=PREREQ
+```
+
+The Instructions to manually set up the connector can be found in the corresponding [Blog Post](https://aws.amazon.com).
```
Comment (baldawar) [inline cdk/README.md:38]: Reminder to update the link here too

File: cdk/README.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+## Deployment
+
+There are 2 deployment options for the CDK stack:
+
+To deploy everything mentioned above, inlcuding the Connector itself please run the following command in this directory:
+```shell
+cdk deploy --context deploymentMode=FULL
+```
+
+To deploy only the Prerequisites and later deploy the connector manually. Please run the following command in this directory:
```
Comment (baldawar) [inline cdk/README.md:32]: It would be helpful to add a line on when PREREQ will be preferred option.

File: cdk/customResources/connectorPlugin/index.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+#
+
+import boto3
+import time
+
+client = boto3.client('kafkaconnect')
+
+
+def on_event(event, context):
+    print(event)
```
Comment (baldawar) [inline cdk/customResources/connectorPlugin/index.py:14]: curious, why not https://docs.python.org/3/library/logging.html ?

File: cdk/customResources/connectorPlugin/index.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    request_type = event['RequestType']
+    if request_type == 'Create': return on_create(event)
+    if request_type == 'Update': return on_update(event)
+    if request_type == 'Delete': return on_delete(event)
+
+    return
```
Comment (baldawar) [inline cdk/customResources/connectorPlugin/index.py:20]: the paranoid engineer in me would prefer to throw an error (or at least log) if an unhandled request type showed up here.

File: cdk/customResources/connectorPlugin/index.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        'PhysicalResourceId': physical_id,
+        'Data': {
+            'arn': response['customPluginArn'],
+            'revision': response['revision']
+        }
+    }
+
+
+def on_update(event):
+    pass
```
Comment (baldawar) [inline cdk/customResources/connectorPlugin/index.py:51]: as this is not typical behaviour, a comment would be helpful.

File: cdk/customResources/connectorPlugin/index.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    print("delete resource with physical id %s" % physical_id)
+    connector_active = True
+    while connector_active:
+        try:
+            response = client.delete_custom_plugin(
+                customPluginArn=physical_id
+            )
+            connector_active = False
+        except:
+            time.sleep(10000)
```
Comment (baldawar) [inline cdk/customResources/connectorPlugin/index.py:65]: I usually prefer to add a comment that this is milliseconds

File: cdk/customResources/connectorPlugin/index.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+def on_update(event):
+    pass
+
+
+def on_delete(event):
+    physical_id = event["PhysicalResourceId"]
+    print("delete resource with physical id %s" % physical_id)
+    connector_active = True
+    while connector_active:
```
Comment (baldawar) [inline cdk/customResources/connectorPlugin/index.py:58]: I'd recommend adding a max retry counter to avoid being stuck in this loop forever. That could get expensive.

File: cdk/transactionAnalyzer/pom.xml [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    <packaging>jar</packaging>
+    <name>kafka-eventbridge-sink-transactionAnalyzer</name>
+    <url>https://github.com/awslabs/eventbridge-kafka-connector</url>
+    <description>Kafka consumer to process financial transactions</description>
+
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <aws.schema.registry.version>1.1.15</aws.schema.registry.version>
+        <maven.compiler.source>11</maven.compiler.source>
+        <maven.compiler.target>11</maven.compiler.target>
```
Comment (baldawar) [inline cdk/transactionAnalyzer/pom.xml:16]: Why not java 17 ?

**Review by baldawar — COMMENTED** [review]
I haven't gone through all files yet. Leaving few comments in case they're helpful.

File: cdk/transactionAnalyzer/src/main/java/software/amazon/event/kafkaconnector/TransactionAnalyzer.java [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+    }
+
+    /**
+     * @param userId The ID of the user in String format
+     * @param value  The transaction record in GenericRecord format
+     * @param notificationSchema The schema for the notification record
+     * @return {@link GenericRecord} A record with the provided schema
+     */
+    private static GenericRecord createNotificationRecord(String userId, GenericRecord value,  Schema notificationSchema) {
+        log.debug("Creating notfication record for");
```
Comment (embano1) [inline cdk/transactionAnalyzer/src/main/java/software/amazon/event/kafkaconnector/TransactionAnalyzer.java:152]: for...missing?

File: cdk/transactionAnalyzer/src/main/java/software/amazon/event/kafkaconnector/TransactionAnalyzer.java [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+        StreamsBuilder builder = new StreamsBuilder();
+        final KStream<String, GenericRecord> events = builder.stream(sourceTopic);
+
+        events
+                .filter((key, value) -> Double.parseDouble(value.get("total").toString()) > 195.00)
+                .map((userId, value) -> {
+                    var notification = createNotificationRecord(userId, value, notificationSchema);
+                    //As this is synthetic test data we can log any value as it does not contain PII
+                    log.info("Detected suspicious transaction. Sending notification to topic. UserId: {} Value: {}", userId, value.get("total").toString());
+                    return KeyValue.pair(UUID.randomUUID().toString(), notification);
```
Comment (embano1) [inline cdk/transactionAnalyzer/src/main/java/software/amazon/event/kafkaconnector/TransactionAnalyzer.java:68]: questions:

- why a random UUID, i.e., not reuse user id?
- what's this key used for? topic key?

File: cdk/transactionAnalyzer/src/main/java/software/amazon/event/kafkaconnector/TransactionAnalyzer.java [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+
+        StreamsBuilder builder = new StreamsBuilder();
+        final KStream<String, GenericRecord> events = builder.stream(sourceTopic);
+
+        events
+                .filter((key, value) -> Double.parseDouble(value.get("total").toString()) > 195.00)
+                .map((userId, value) -> {
+                    var notification = createNotificationRecord(userId, value, notificationSchema);
+                    //As this is synthetic test data we can log any value as it does not contain PII
+                    log.info("Detected suspicious transaction. Sending notification to topic. UserId: {} Value: {}", userId, value.get("total").toString());
```
Comment (embano1) [inline cdk/transactionAnalyzer/src/main/java/software/amazon/event/kafkaconnector/TransactionAnalyzer.java:67]: nit:

```suggestion
                    log.info("Detected suspicious transaction. Sending notification to topic {}. UserId: {} Value: {}", notificationsTopic, userId, value.get("total").toString());
```

File: cdk/transactionAnalyzer/src/main/java/software/amazon/event/kafkaconnector/TransactionAnalyzer.java [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+        var topicCreator = new TransactionAnalyzerTopicCreator(properties);
+        topicCreator.createTopic(notificationsTopic, Integer.valueOf(partitionCount), Short.valueOf(replicationFactor));
+        topicCreator.createTopic(sourceTopic, Integer.valueOf(partitionCount), Short.valueOf(replicationFactor));
+        topicCreator.close();
+
+        StreamsBuilder builder = new StreamsBuilder();
+        final KStream<String, GenericRecord> events = builder.stream(sourceTopic);
+
+        events
+                .filter((key, value) -> Double.parseDouble(value.get("total").toString()) > 195.00)
```
Comment (embano1) [inline cdk/transactionAnalyzer/src/main/java/software/amazon/event/kafkaconnector/TransactionAnalyzer.java:63]: nit: would be nice to have a time-windowed filter [1] instead of a primitive one like this because one could argue this can be done directly in EventBridge too, i.e., without going through a notification topic.

[1]
Say, `> 195.00 within 3 minutes`

File: cdk/transactionAnalyzer/src/main/java/software/amazon/event/kafkaconnector/TransactionAnalyzer.java [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+            }
+        });
+
+        try {
+            log.info("Starting Transaction Analyzer ...");
+            streams.start();
+            latch.await();
+
+        } catch (Throwable e) {
+            log.info("Transaction Analyzer failed. Shutting down application.");
```
Comment (embano1) [inline cdk/transactionAnalyzer/src/main/java/software/amazon/event/kafkaconnector/TransactionAnalyzer.java:91]: nit: log exception?

File: cdk/transactionAnalyzer/src/main/java/software/amazon/event/kafkaconnector/TransactionAnalyzerTopicCreator.java [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+        var newTopic = new NewTopic(topicName, partitionCount, replicationFactor);
+        try {
+            var result = admin.createTopics(Collections.singleton(newTopic));
+            result.all().get(30, TimeUnit.SECONDS);
+            log.info("Topic created: " + topicName);
+        } catch (ExecutionException | InterruptedException | TimeoutException e) {
+            if (e.getCause() != null && e.getCause() instanceof TopicExistsException) {
+                log.info("Topic exists: {}", newTopic);
+            } else {
+                log.error("Topic create failed with: " + e);
```
Comment (embano1) [inline cdk/transactionAnalyzer/src/main/java/software/amazon/event/kafkaconnector/TransactionAnalyzerTopicCreator.java:51]: nit: if topic create fails, this logs only but what would the caller do since there's no exception (re)thrown?

File: pom.xml [hunk 18]
```diff
@@ -227,6 +227,7 @@
                                 <exclude>**/THIRD-PARTY-LICENSES</exclude>
                                 <exclude>src/test/resources/**</exclude>
                                 <exclude>src/main/resources/**</exclude>
+                                <exclude>cdk/**</exclude>
```
Comment (embano1) [inline pom.xml:230]: 👍🏻

File: cdk/README.md [hunk 19]
```diff
@@ -0,0 +1,46 @@
+# Kafka Connector for Amazon EventBridge Deployment on AWS
+
+This CDK stack is sample deployment to for the Kafka Connector for Amazon EventBridge. 
+
+> :warning: **This sample is not meant to be deployed to production environments!**
```
Comment (embano1) [inline cdk/README.md:5]: Better highlighting

```suggestion
> **Warning**  
> This sample is not meant to be deployed to production environments!**
```

File: cdk/README.md [hunk 20]
```diff
@@ -0,0 +1,46 @@
+# Kafka Connector for Amazon EventBridge Deployment on AWS
+
+This CDK stack is sample deployment to for the Kafka Connector for Amazon EventBridge.
```
Comment (embano1) [inline cdk/README.md:3]: "to for"

File: cdk/README.md (hunk 3 above)
Comment (embano1) [reply cdk/README.md:19]: IMHO this is a placeholder for the final link

Comment (embano1) [reply cdk/README.md:19]: Btw: the link should be highlighted in the top section of this README to showcase this is the accompanying resource of the official blog post.

File: cdk/connectorConfigurationJson.json [hunk 21]
```diff
@@ -0,0 +1,14 @@
+{
+  "connector.class": "software.amazon.event.kafkaconnector.EventBridgeSinkConnector",
+  "topics": "notifications",
+  "aws.eventbridge.eventbus.arn": "arn:aws:events:us-east-1:00000000000:event-bus/eventbridge-sink-eventbus",
```
Comment (embano1) [inline cdk/connectorConfigurationJson.json:4]: assuming account ID will be replaced by the deployment script?

File: cdk/customResources/connectorPlugin/index.py (hunk 7 above)
Comment (embano1) [reply cdk/customResources/connectorPlugin/index.py:20]: +1

switch/case/default :)

File: cdk/README.md (hunk 3 above)
Comment (flo-mair) [reply cdk/README.md:19]: Yes this will link to the actual blog post we are publishing

File: README.md (hunk 1 above)
Comment (flo-mair) [reply README.md:100]: Github has better JSON syntax highlighting when specifying yaml. Please don't ask me why :)

File: cdk/cdk.context.json (hunk 2 above)
Comment (flo-mair) [reply cdk/cdk.context.json:3]: CDK warning surpression

File: cdk/transactionAnalyzer/src/main/java/software/amazon/event/kafkaconnector/TransactionAnalyzer.java (hunk 13 above)
Comment (flo-mair) [reply cdk/transactionAnalyzer/src/main/java/software/amazon/event/kafkaconnector/TransactionAnalyzer.java:68]: userId is already part of notification, using UUID to avoid any potential hot partitions. Didn't want to set it to null, so users can see how keys are sent to EventBridge.

File: cdk/transactionAnalyzer/src/main/java/software/amazon/event/kafkaconnector/TransactionAnalyzer.java (hunk 15 above)
Comment (flo-mair) [reply cdk/transactionAnalyzer/src/main/java/software/amazon/event/kafkaconnector/TransactionAnalyzer.java:63]: This is not possible atm. Ping me offline to chat about why.

File: cdk/connectorConfigurationJson.json (hunk 21 above)
Comment (flo-mair) [reply cdk/connectorConfigurationJson.json:4]: Yes, same for region and event bus name

File: cdk/customResources/connectorPlugin/index.py (hunk 7 above)
Comment (flo-mair) [reply cdk/customResources/connectorPlugin/index.py:20]: let me check if lambda has python 3.10 runtime available. Python was a bit late to the switch case party.

File: README.md (hunk 1 above)
Comment (embano1) [reply README.md:100]: image

Above uses `json5` below uses `yaml`

Comment (embano1) [reply README.md:100]: IMHO what you're wondering is why there's no syntax highlighting with `json5` - which is weird, but allows comments in JSON objects. Feel free to change the examples without comments to `json` only for formatting.

File: cdk/transactionAnalyzer/pom.xml (hunk 11 above)
Comment (embano1) [reply cdk/transactionAnalyzer/pom.xml:16]: The repo uses `11` too, so IMHO it's ok to keep it consistent.

File: cdk/README.md [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+- Amazon S3 Bucket
+- IAM Roles & Security Groups
+- VPC & related components
+- Custom Amazon EventBridge Bus
+- Amazon ECS cluster with 2 Services (Data Producer & Transaction Analyzer)
+- (Optional) Amazon MSK Connector
+
+To learn more about those components please read the blog on the [AWS Big Data Blog](https://aws.amazon.com).
+
+![Architecture Diagram](./images/architecture.png)
```
Comment (embano1) [inline cdk/README.md:21]: nit: on the architecture diagram, the arrow from EventBridge should be inverse, i.e., pointing to the targets because of the push-based invocation (no "consume")

File: cdk/lib/kafkaEventBridgeSinkStack.ts [hunk 23]
```diff
@@ -0,0 +1,203 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (embano1) [inline cdk/lib/kafkaEventBridgeSinkStack.ts:201]: nit: remove

File: cdk/transactionAnalyzer/src/main/java/software/amazon/event/kafkaconnector/TransactionAnalyzer.java [hunk 24]
```diff
@@ -0,0 +1,164 @@
...
+        var replicationFactor = System.getenv().getOrDefault("REPLICATION_FACTOR", "3");
+        var partitionCount = System.getenv().getOrDefault("PARTITION_COUNT", "10");
+        var notificationSchemaFile = System.getenv().getOrDefault("NOTIFICATION_SCHEMA_FILE", "notification.avsc");
+
+        var notificationSchema = createSchema(notificationSchemaFile);
+        var properties = setConfiguration();
+
+        var topicCreator = new TransactionAnalyzerTopicCreator(properties);
+        topicCreator.createTopic(notificationsTopic, Integer.valueOf(partitionCount), Short.valueOf(replicationFactor));
+        topicCreator.createTopic(sourceTopic, Integer.valueOf(partitionCount), Short.valueOf(replicationFactor));
```
Comment (embano1) [inline cdk/transactionAnalyzer/src/main/java/software/amazon/event/kafkaconnector/TransactionAnalyzer.java:56]: Shouldn't `sourceTopic` already exist? Or is this to prevent races between producer/consumer startup?

File: cdk/lib/kafkaEventBridgeSinkStack.ts [hunk 25]
```diff
@@ -0,0 +1,203 @@
...
+            resources: [`arn:aws:kafka:${this.region}:${this.account}:cluster/${cluster.clusterName}/*`]
+        }))
+
+        connectorRole.addToPolicy(new PolicyStatement({
+            actions: [
+                "kafka-cluster:*Topic*",
+                "kafka-cluster:WriteData",
+                "kafka-cluster:ReadData"
+            ],
+            resources: [`arn:aws:kafka:${this.region}:${this.account}:topic/${cluster.clusterName}/*`]
```
Comment (embano1) [inline cdk/lib/kafkaEventBridgeSinkStack.ts:180]: nit: if topic name would be a property used throughout this demo (producer, consumer, IAM), this could be scoped down further to prevent broad IAM access

File: cdk/kafkaProducer/src/main/java/software/amazon/event/kafkaconnector/ProducerMetricReporter.java [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @Override
+    protected void runOneIteration() {
+
+        for (Map.Entry<MetricName, ? extends Metric> entry : kafkaProducer.metrics().entrySet()) {
+            if (Objects.equals(entry.getKey().description(), "The total number of records sent.")) {
+                log.info(entry.getKey().description() + " " + entry.getValue().metricValue().toString());
+            }
+        }
```
Comment (baldawar) [inline cdk/kafkaProducer/src/main/java/software/amazon/event/kafkaconnector/ProducerMetricReporter.java:44]: This looks to be doing a running sum as there's no update to the metric-value after being read. This conflicts with the comment on L22 which says the class will report how many records in internal.

File: cdk/lib/kafkaEventBridgeSinkStack.ts [hunk 27]
```diff
@@ -0,0 +1,203 @@
...
+            })
+        }
+
+        connectorRole.addToPolicy(new iam.PolicyStatement({
+            actions: ['events:PutEvents'],
+            resources: [`arn:aws:events:${this.region}:${this.account}:event-bus/eventbridge-sink-eventbus`]
+        }))
+
+        connectorRole.addToPolicy(new iam.PolicyStatement({
+            actions: ['glue:GetSchemaVersion'],
```
Comment (embano1) [inline cdk/lib/kafkaEventBridgeSinkStack.ts:161]: Question: is this sufficient to retrieve the corresponding schema? Don't we need [GetSchema](https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api-schema-registry-api.html#aws-glue-api-schema-registry-api-GetSchema)?

File: cdk/lib/analyzerConstruct.ts [hunk 28]
```diff
@@ -83,6 +86,7 @@ export class Analyzer extends Construct {
                 "kafka-cluster:ReadData"
             ],
             resources: [`arn:aws:kafka:${props.region}:${props.account}:topic/${props.clusterName}/*`]
+            //Topic scope can not be limited as user defined topics are allowed
```
Comment (embano1) [inline cdk/lib/analyzerConstruct.ts:89]: nit: typically comments are above the code line

**Comment by embano1** [comment]
Please squash the commits following conventional commits patterns

File: README.md (hunk 1 above)
Comment (flo-mair) [reply README.md]: Done

File: cdk/README.md (hunk 5 above)
Comment (flo-mair) [reply cdk/README.md:33]: Done
